File: backend/src/routes/shifts.js
const express = require('express');
const { body, query, validationResult } = require('express-validator');
const { Shift } = require('../models/sales');
const { authMiddleware, employeeAuthMiddleware } = require('../middleware/auth');
const { AuditLog } = require('../models/audit');
const logger = require('../utils/logger');

const router = express.Router();

// Open shift
router.post('/open', [
  authMiddleware,
  employeeAuthMiddleware,
  body('cash_open').isNumeric().withMessage('Cash open amount is required')
], async (req, res, next) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        error: 'E_VALIDATION',
        message: 'Validation failed',
        details: errors.array()
      });
    }

    const { cash_open, notes } = req.body;

    // Check if there's already an open shift for this device
    const existingShift = await Shift.findOne({
      store_id: req.store_id,
      device_id: req.device_id,
      status: 'open'
    });

    if (existingShift) {
      return res.status(400).json({
        error: 'E_SHIFT_ALREADY_OPEN',
        message: 'Shift is already open for this device'
      });
    }

    // Create new shift
    const shift = new Shift({
      store_id: req.store_id,
      device_id: req.device_id,
      opened_by: req.employee_id,
      cash_open: cash_open,
      notes: notes
    });

    await shift.save();

    // Log shift opening
    await AuditLog.logEvent({
      event_type: 'shift.opened',
      actor_id: req.employee_id,
      device_id: req.device_id,
      payload: {
        shift_id: shift.shift_id,
        cash_open: cash_open,
        notes: notes
      },
      ip_address: req.ip,
      store_id: req.store_id,
      tenant_id: req.tenant_id
    });

    logger.info(`Shift opened: ${shift.shift_id}`, {
      shift_id: shift.shift_id,
      device_id: req.device_id,
      employee_id: req.employee_id,
      cash_open: cash_open
    });

    res.status(201).json({
      shift_id: shift.shift_id,
      store_id: shift.store_id,
      device_id: shift.device_id,
      opened_by: shift.opened_by,
      open_at: shift.open_at,
      cash_open: shift.cash_open,
      status: shift.status,
      totals: shift.totals
    });

  } catch (error) {
    next(error);
  }
});

// Close shift
router.post('/close', [
  authMiddleware,
  employeeAuthMiddleware,
  body('cash_counted').isNumeric().withMessage('Cash counted amount is required')
], async (req, res, next) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        error: 'E_VALIDATION',
        message: 'Validation failed',
        details: errors.array()
      });
    }

    const { cash_counted, notes } = req.body;

    // Find open shift
    const shift = await Shift.findOne({
      store_id: req.store_id,
      device_id: req.device_id,
      status: 'open'
    });

    if (!shift) {
      return res.status(400).json({
        error: 'E_NO_OPEN_SHIFT',
        message: 'No open shift found for this device'
      });
    }

    // Close shift
    shift.closeShift(req.employee_id, cash_counted);
    if (notes) shift.notes = notes;
    await shift.save();

    // Log shift closing
    await AuditLog.logEvent({
      event_type: 'shift.closed',
      actor_id: req.employee_id,
      device_id: req.device_id,
      payload: {
        shift_id: shift.shift_id,
        cash_open: shift.cash_open,
        cash_counted: cash_counted,
        cash_expected: shift.cash_expected,
        cash_diff: shift.cash_diff,
        totals: shift.totals,
        notes: notes
      },
      ip_address: req.ip,
      store_id: req.store_id,
      tenant_id: req.tenant_id
    });

    logger.info(`Shift closed: ${shift.shift_id}`, {
      shift_id: shift.shift_id,
      device_id: req.device_id,
      employee_id: req.employee_id,
      cash_diff: shift.cash_diff
    });

    res.json({
      shift_id: shift.shift_id,
      store_id: shift.store_id,
      device_id: shift.device_id,
      opened_by: shift.opened_by,
      closed_by: shift.closed_by,
      open_at: shift.open_at,
      close_at: shift.close_at,
      cash_open: shift.cash_open,
      cash_counted: shift.cash_counted,
      cash_expected: shift.cash_expected,
      cash_diff: shift.cash_diff,
      status: shift.status,
      totals: shift.totals,
      notes: shift.notes
    });

  } catch (error) {
    next(error);
  }
});

// Get current shift
router.get('/current', [
  authMiddleware,
  employeeAuthMiddleware
], async (req, res, next) => {
  try {
    const shift = await Shift.findOne({
      store_id: req.store_id,
      device_id: req.device_id,
      status: 'open'
    });

    if (!shift) {
      return res.status(404).json({
        error: 'E_NO_OPEN_SHIFT',
        message: 'No open shift found for this device'
      });
    }

    res.json({
      shift_id: shift.shift_id,
      store_id: shift.store_id,
      device_id: shift.device_id,
      opened_by: shift.opened_by,
      open_at: shift.open_at,
      cash_open: shift.cash_open,
      status: shift.status,
      totals: shift.totals,
      notes: shift.notes
    });

  } catch (error) {
    next(error);
  }
});

// Get shifts list
router.get('/', [
  authMiddleware,
  employeeAuthMiddleware,
  query('limit').optional().isInt({ min: 1, max: 100 }),
  query('offset').optional().isInt({ min: 0 }),
  query('from_date').optional().isISO8601(),
  query('to_date').optional().isISO8601(),
  query('status').optional().isIn(['open', 'closed'])
], async (req, res, next) => {
  try {
    const errors = validationResult(req);
    if (!errors.isEmpty()) {
      return res.status(400).json({
        error: 'E_VALIDATION',
        message: 'Validation failed',
        details: errors.array()
      });
    }

    const { limit = 50, offset = 0, from_date, to_date, status } = req.query;

    // Build query
    const query = {
      store_id: req.store_id
    };

    if (status) query.status = status;

    if (from_date || to_date) {
      query.open_at = {};
      if (from_date) query.open_at.$gte = new Date(from_date);
      if (to_date) query.open_at.$lte = new Date(to_date);
    }

    const shifts = await Shift.find(query)
      .sort({ open_at: -1 })
      .limit(parseInt(limit))
      .skip(parseInt(offset));

    const total = await Shift.countDocuments(query);

    res.json({
      shifts: shifts.map(shift => ({
        shift_id: shift.shift_id,
        device_id: shift.device_id,
        opened_by: shift.opened_by,
        closed_by: shift.closed_by,
        open_at: shift.open_at,
        close_at: shift.close_at,
        cash_open: shift.cash_open,
        cash_counted: shift.cash_counted,
        cash_expected: shift.cash_expected,
        cash_diff: shift.cash_diff,
        status: shift.status,
        totals: shift.totals
      })),
      pagination: {
        total: total,
        limit: parseInt(limit),
        offset: parseInt(offset),
        has_more: total > parseInt(offset) + parseInt(limit)
      }
    });

  } catch (error) {
    next(error);
  }
});

// Get shift by ID
router.get('/:shift_id', [
  authMiddleware,
  employeeAuthMiddleware
], async (req, res, next) => {
  try {
    const { shift_id } = req.params;

    const shift = await Shift.findOne({
      shift_id: shift_id,
      store_id: req.store_id
    });

    if (!shift) {
      return res.status(404).json({
        error: 'E_SHIFT_NOT_FOUND',
        message: 'Shift not found'
      });
    }

    res.json({
      shift_id: shift.shift_id,
      store_id: shift.store_id,
      device_id: shift.device_id,
      opened_by: shift.opened_by,
      closed_by: shift.closed_by,
      open_at: shift.open_at,
      close_at: shift.close_at,
      cash_open: shift.cash_open,
      cash_counted: shift.cash_counted,
      cash_expected: shift.cash_expected,
      cash_diff: shift.cash_diff,
      status: shift.status,
      totals: shift.totals,
      notes: shift.notes,
      created_at: shift.created_at,
      updated_at: shift.updated_at
    });

  } catch (error) {
    next(error);
  }
});

module.exports = router;

File: backend-fastapi/app/services/shifts.py
"""Shift Service - Complete Implementation"""
from typing import Dict, Any, Optional
from datetime import datetime, timedelta
from decimal import Decimal

from app.models.shifts import Shift, ShiftOpenRequest, ShiftCloseRequest
from app.repositories.shifts import ShiftRepository
from app.utils.logging import LoggerMixin
from app.utils.errors import PlayParkException, ErrorCode


class ShiftService(LoggerMixin):
    def __init__(self, shift_repo: ShiftRepository):
        self.shift_repo = shift_repo
    
    async def open_shift(self, request: ShiftOpenRequest, user) -> Dict[str, Any]:
        """Open a new shift"""
        self.logger.info("Opening shift", user_id=user.employee_id, store_id=user.store_id)
        
        try:
            # Check if there's already an open shift for this user
            existing_shift = await self.shift_repo.get_current_shift_by_employee(user.employee_id)
            if existing_shift and existing_shift.status == "open":
                raise PlayParkException(
                    error_code=ErrorCode.BAD_REQUEST,
                    message="Employee already has an open shift"
                )
            
            # Create new shift
            shift = Shift(
                shift_id=f"shift_{datetime.utcnow().strftime('%Y%m%d%H%M%S')}_{user.employee_id}",
                tenant_id=user.tenant_id if hasattr(user, 'tenant_id') else "default",
                store_id=user.store_id,
                device_id=user.device_id if hasattr(user, 'device_id') else "default",
                opened_by=user.employee_id,
                open_at=datetime.utcnow(),
                cash_open=request.cash_open,
                status="open",
                notes=request.notes if hasattr(request, 'notes') else None
            )
            
            created_shift = await self.shift_repo.create_shift(shift)
            
            return {
                "shift_id": created_shift.shift_id,
                "store_id": created_shift.store_id,
                "employee_id": created_shift.opened_by,
                "status": created_shift.status,
                "opening_cash": created_shift.cash_open,
                "opening_time": created_shift.open_at,
                "notes": created_shift.notes
            }
            
        except PlayParkException:
            raise
        except Exception as e:
            self.logger.error("Error opening shift", error=str(e))
            raise PlayParkException(
                error_code=ErrorCode.INTERNAL_ERROR,
                message="Failed to open shift"
            )
    
    async def close_shift(self, request: ShiftCloseRequest, user) -> Dict[str, Any]:
        """Close the current shift"""
        self.logger.info("Closing shift", user_id=user.employee_id)
        
        try:
            # Get current shift
            current_shift = await self.shift_repo.get_current_shift_by_employee(user.employee_id)
            if not current_shift or current_shift.status != "open":
                raise PlayParkException(
                    error_code=ErrorCode.BAD_REQUEST,
                    message="No open shift found for employee"
                )
            
            # Update shift with closing information
            closing_time = datetime.utcnow()
            shift_duration = closing_time - current_shift.open_at
            
            # Calculate totals (in real app, you'd get these from sales)
            total_sales = Decimal('0')  # Placeholder
            total_refunds = Decimal('0')  # Placeholder
            closing_cash = request.cash_counted or Decimal('0')
            cash_difference = closing_cash - current_shift.cash_open - total_sales + total_refunds
            
            updated_shift = await self.shift_repo.update_shift(
                current_shift.shift_id,
                {
                    "status": "closed",
                    "close_at": closing_time,
                    "closed_by": user.employee_id,
                    "cash_counted": float(closing_cash),
                    "cash_expected": float(current_shift.cash_open + total_sales - total_refunds),
                    "cash_diff": float(cash_difference),
                    "notes": request.notes
                }
            )
            
            return {
                "shift_id": updated_shift.shift_id,
                "status": updated_shift.status,
                "opening_time": updated_shift.open_at,
                "closing_time": updated_shift.close_at,
                "duration_minutes": int((updated_shift.close_at - updated_shift.open_at).total_seconds() / 60),
                "opening_cash": updated_shift.cash_open,
                "closing_cash": updated_shift.cash_counted,
                "total_sales": float(total_sales),
                "cash_difference": updated_shift.cash_diff,
                "notes": updated_shift.notes
            }
            
        except PlayParkException:
            raise
        except Exception as e:
            self.logger.error("Error closing shift", error=str(e))
            raise PlayParkException(
                error_code=ErrorCode.INTERNAL_ERROR,
                message="Failed to close shift"
            )
    
    async def get_current_shift(self, user) -> Optional[Dict[str, Any]]:
        """Get the current open shift for the user"""
        self.logger.info("Getting current shift", user_id=user.employee_id)
        
        try:
            shift = await self.shift_repo.get_current_shift_by_employee(user.employee_id)
            
            if not shift or shift.status != "open":
                return None
            
            return {
                "shift_id": shift.shift_id,
                "store_id": shift.store_id,
                "employee_id": shift.opened_by,
                "status": shift.status,
                "opening_cash": shift.cash_open,
                "opening_time": shift.open_at,
                "notes": shift.notes,
                "duration_minutes": int((datetime.utcnow() - shift.open_at).total_seconds() / 60)
            }
            
        except Exception as e:
            self.logger.error("Error getting current shift", error=str(e))
            raise PlayParkException(
                error_code=ErrorCode.INTERNAL_ERROR,
                message="Failed to retrieve current shift"
            )
